File: apps/sveltekit/board/src/lib/tickets.ts
import fs from "node:fs";
import path from "node:path";
import YAML from "yaml";

export const STATUSES = [
	"blocked",
	"design-needed",
	"plan-needed",
	"ready",
	"in-review",
	"done",
] as const;

export type Status = (typeof STATUSES)[number];

export const STATUS_LABELS: Record<Status, string> = {
	blocked: "Blocked",
	"design-needed": "Design needed",
	"plan-needed": "Plan needed",
	ready: "Ready",
	"in-review": "In review",
	done: "Done",
};

export interface Ticket {
	id: string;
	title: string;
	status: Status;
	dependsOn: string[];
	body: string;
	filePath: string;
	epic: string | null;
}

export interface EpicGroup {
	epic: string | null;
	tickets: Ticket[];
}

export function deriveEpic(filePath: string, ticketsDirectory: string): string | null {
	const relative = path.relative(ticketsDirectory, filePath);
	const segments = relative.split(path.sep);
	return segments.length > 1 ? (segments[0] ?? null) : null;
}

export function collectEpics(tickets: Ticket[]): string[] {
	const epics = new Set<string>();
	for (const ticket of tickets) {
		if (ticket.epic !== null) {
			epics.add(ticket.epic);
		}
	}
	return [...epics].sort();
}

export function groupByEpic(tickets: Ticket[]): EpicGroup[] {
	const groups = new Map<string | null, Ticket[]>();
	for (const ticket of tickets) {
		const existing = groups.get(ticket.epic);
		if (existing) {
			existing.push(ticket);
		} else {
			groups.set(ticket.epic, [ticket]);
		}
	}
	const result: EpicGroup[] = [];
	for (const [epic, epicTickets] of groups) {
		result.push({ epic, tickets: epicTickets });
	}
	result.sort((a, b) => {
		if (a.epic === null && b.epic === null) return 0;
		if (a.epic === null) return 1;
		if (b.epic === null) return -1;
		return a.epic.localeCompare(b.epic);
	});
	return result;
}

export function parseFrontmatter(text: string): {
	metadata: Record<string, unknown>;
	body: string;
} {
	if (!text.startsWith("---\n")) {
		return { metadata: {}, body: text };
	}
	const endIndex = text.indexOf("\n---\n", 4);
	if (endIndex === -1) {
		return { metadata: {}, body: text };
	}
	const header = text.slice(4, endIndex);
	const body = text.slice(endIndex + 5);
	return { metadata: YAML.parse(header) as Record<string, unknown>, body };
}

export function dumpFrontmatter(
	metadata: Record<string, unknown>,
	body: string,
): string {
	const header = YAML.stringify(metadata).trimEnd();
	return `---\n${header}\n---\n${body}`;
}

export function loadTicket(filePath: string): Ticket {
	const text = fs.readFileSync(filePath, "utf-8");
	const { metadata, body } = parseFrontmatter(text);
	const dependsOn = metadata["depends_on"];
	return {
		id: metadata["id"] as string,
		title: metadata["title"] as string,
		status: metadata["status"] as Status,
		dependsOn: Array.isArray(dependsOn) ? (dependsOn as string[]) : [],
		body,
		filePath,
		epic: null,
	};
}

function ticketSortKey(filePath: string): number {
	const name = path.basename(filePath, ".md");
	const prefix = name.split("-")[0] ?? "";
	const numeric = prefix.slice(1);
	return /^\d+$/.test(numeric) ? parseInt(numeric, 10) : 0;
}

function findTicketFiles(directory: string): string[] {
	if (!fs.existsSync(directory)) {
		return [];
	}
	const results: string[] = [];
	for (const entry of fs.readdirSync(directory, { withFileTypes: true })) {
		if (entry.isDirectory()) {
			results.push(...findTicketFiles(path.join(directory, entry.name)));
		} else if (/^t\d+.*\.md$/.test(entry.name)) {
			results.push(path.join(directory, entry.name));
		}
	}
	return results;
}

export function loadTickets(directory: string): Ticket[] {
	const files = findTicketFiles(directory).sort(
		(a: string, b: string) => ticketSortKey(a) - ticketSortKey(b),
	);

	const tickets: Ticket[] = [];
	for (const filePath of files) {
		const text = fs.readFileSync(filePath, "utf-8");
		if (!text.startsWith("---\n")) {
			continue;
		}
		const ticket = loadTicket(filePath);
		ticket.epic = deriveEpic(filePath, directory);
		tickets.push(ticket);
	}
	return tickets;
}

export function ticketsByStatus(
	tickets: Ticket[],
): Record<Status, Ticket[]> {
	const grouped: Record<Status, Ticket[]> = {
		blocked: [],
		"design-needed": [],
		"plan-needed": [],
		ready: [],
		"in-review": [],
		done: [],
	};
	for (const ticket of tickets) {
		const bucket = grouped[ticket.status];
		if (bucket) {
			bucket.push(ticket);
		}
	}
	return grouped;
}

export function updateTicketStatus(
	ticketId: string,
	newStatus: Status | string,
	directory: string,
): void {
	for (const filePath of findTicketFiles(directory)) {
		const text = fs.readFileSync(filePath, "utf-8");
		if (!text.startsWith("---\n")) {
			continue;
		}
		const { metadata, body } = parseFrontmatter(text);
		if (metadata["id"] === ticketId) {
			metadata["status"] = newStatus;
			fs.writeFileSync(filePath, dumpFrontmatter(metadata, body), "utf-8");
			return;
		}
	}
	throw new Error(`Ticket ${ticketId} not found`);
}

File: scripts/src/scripts/tickets.py
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import yaml

TICKETS_DIRECTORY = Path(__file__).resolve().parent.parent.parent.parent / "agent" / "tickets"

STATUSES = ["blocked", "design-needed", "plan-needed", "ready", "done"]


@dataclass
class Ticket:
    id: str
    title: str
    status: str
    depends_on: list[str]
    body: str
    file_path: Path


def parse_frontmatter(text: str) -> tuple[dict[str, Any], str]:
    if not text.startswith("---\n"):
        return {}, text
    end = text.index("\n---\n", 4)
    header = text[4:end]
    body = text[end + 5 :]
    return yaml.safe_load(header), body


def dump_frontmatter(metadata: dict[str, Any], body: str) -> str:
    header = yaml.dump(metadata, default_flow_style=False, sort_keys=False).rstrip("\n")
    return f"---\n{header}\n---\n{body}"


def load_ticket(path: Path) -> Ticket:
    text = path.read_text(encoding="utf-8")
    metadata, body = parse_frontmatter(text)
    return Ticket(
        id=metadata["id"],
        title=metadata["title"],
        status=metadata["status"],
        depends_on=metadata.get("depends_on", []),
        body=body,
        file_path=path,
    )


def load_tickets(directory: Path | None = None) -> list[Ticket]:
    directory = directory or TICKETS_DIRECTORY
    tickets = []
    for path in sorted(directory.glob("t*.md"), key=_ticket_sort_key):
        text = path.read_text(encoding="utf-8")
        if not text.startswith("---\n"):
            continue
        tickets.append(load_ticket(path))
    return tickets


def _ticket_sort_key(path: Path) -> int:
    name = path.stem
    prefix = name.split("-")[0]
    return int(prefix[1:]) if prefix[1:].isdigit() else 0


def tickets_by_status(tickets: list[Ticket]) -> dict[str, list[Ticket]]:
    grouped: dict[str, list[Ticket]] = {status: [] for status in STATUSES}
    for ticket in tickets:
        grouped.setdefault(ticket.status, []).append(ticket)
    return grouped


def update_ticket_status(ticket_id: str, new_status: str, directory: Path | None = None) -> None:
    directory = directory or TICKETS_DIRECTORY
    for path in directory.glob("t*.md"):
        text = path.read_text(encoding="utf-8")
        if not text.startswith("---\n"):
            continue
        metadata, body = parse_frontmatter(text)
        if metadata.get("id") == ticket_id:
            metadata["status"] = new_status
            path.write_text(dump_frontmatter(metadata, body), encoding="utf-8")
            return
    raise ValueError(f"Ticket {ticket_id} not found")
